mm2src/coins/eth.rs

```diff
@@ -1243,55 +1245,348 @@ impl SwapOps for EthCoin {
+        let signed = try_tx_fus!(SignedEthTx::new(tx));
+        let swap_contract_address = try_tx_fus!(input.swap_contract_address.try_to_address());
+
+        Box::new(
+            self.watcher_spend_hash_time_locked_payment(
+                signed,
+                input.secret_hash,
+                swap_contract_address,
+                input.secret,
+                input.taker_pub,
+            )
+            .map(TransactionEnum::from),
+        )
     }
 
     fn create_maker_payment_spend_preimage(
         &self,
-        _maker_payment_tx: &[u8],
+        maker_payment_tx: &[u8],
         _time_lock: u32,
         _maker_pub: &[u8],
         _secret_hash: &[u8],
         _swap_unique_data: &[u8],
     ) -> TransactionFut {
-        unimplemented!();
+        let tx: UnverifiedTransaction = try_tx_fus!(rlp::decode(maker_payment_tx));
+        let signed = try_tx_fus!(SignedEthTx::new(tx));
+        let fut = async move {
+            let tx = TransactionEnum::from(signed);
+            Ok(tx)
+        };
+
+        Box::new(fut.boxed().compat())
     }
 
     fn create_taker_payment_refund_preimage(
         &self,
-        _taker_payment_tx: &[u8],
+        taker_payment_tx: &[u8],
         _time_lock: u32,
         _maker_pub: &[u8],
         _secret_hash: &[u8],
         _swap_contract_address: &Option<BytesJson>,
         _swap_unique_data: &[u8],
     ) -> TransactionFut {
-        unimplemented!();
+        let tx: UnverifiedTransaction = try_tx_fus!(rlp::decode(taker_payment_tx));
+        let signed = try_tx_fus!(SignedEthTx::new(tx));
+        let fut = async move {
+            let tx = TransactionEnum::from(signed);
+            Ok(tx)
+        };
+
+        Box::new(fut.boxed().compat())
     }
 
-    fn send_taker_payment_refund_preimage(&self, _taker_refunds_payment: &[u8]) -> TransactionFut {
-        unimplemented!();
+    fn send_taker_payment_refund_preimage(
+        &self,
+        watcher_refunds_payment_args: SendWatcherRefundsPaymentArgs,
+    ) -> TransactionFut {
+        let tx: UnverifiedTransaction = try_tx_fus!(rlp::decode(watcher_refunds_payment_args.payment_tx));
+        let signed = try_tx_fus!(SignedEthTx::new(tx));
+        let swap_contract_address = try_tx_fus!(watcher_refunds_payment_args.swap_contract_address.try_to_address());
+
+        Box::new(
+            self.watcher_refunds_hash_time_locked_payment(
+                swap_contract_address,
+                signed,
+                watcher_refunds_payment_args.secret_hash,
+                watcher_refunds_payment_args.other_pubkey,
+            )
+            .map(TransactionEnum::from),
+        )
     }
 
-    fn watcher_validate_taker_fee(&self, _input: WatcherValidateTakerFeeInput) -> ValidatePaymentFut<()> {
-        unimplemented!();
+    fn watcher_validate_taker_fee(&self, validate_fee_args: WatcherValidateTakerFeeInput) -> ValidatePaymentFut<()> {
```

Comment:
`validate_fee` takes as parameter a copy of the taker fee, then requests the transaction from the chain using the hash. In the watcher, we don't have a copy of the transaction, we only have the hash. If we use the `validate_fee` method, we'd have to request the transaction from the chain twice: once before calling the method and once in the method.

Reply:
I think for the watcher node it should be a problem to request the transaction twice. But on the other hand, it will simplify the code significantly.
Also as another solution, `EthCoin::validate_fee` method could take an extended `EthValidateFeeArgs`:
```rust
    pub fee_tx_hash: &'a [u8], // This is the only difference between `ValidateFeeArgs` and `EthValidateFeeArgs`.
    pub expected_sender: &'a [u8],
    pub fee_addr: &'a [u8],
    pub amount: &'a BigDecimal,
    pub min_block_number: u64,
    pub uuid: &'a [u8],
```

Reply:
Are you talking about writing another method that takes as parameter `EthValidateFeeArgs`, which is then called both by `EthCoin::validate_fee` and `EthCoin::watcher_validate_taker_fee`?

Reply:
I also added this to the [todo list](https://github.com/KomodoPlatform/atomicDEX-API/issues/1431#issuecomment-1377916727) in the issue

Reply:
Yes, exactly :)
> Are you talking about writing another method that takes as parameter EthValidateFeeArgs, which is then called both by EthCoin::validate_fee and EthCoin::watcher_validate_taker_fee?